org.eclipse.cdt.cross.arc.gnu/src/org/eclipse/cdt/cross/arc/gnu/common/CompilerPluginVersionsChecker.java

```diff
@@ -0,0 +1,103 @@
+package org.eclipse.cdt.cross.arc.gnu.common;
+
+import java.util.HashSet;
+
+import org.eclipse.cdt.make.core.scannerconfig.IScannerInfoCollector;
+import org.eclipse.cdt.make.core.scannerconfig.InfoContext;
+import org.eclipse.cdt.make.internal.core.scannerconfig2.PerProjectSICollector;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.ProjectScope;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences;
+import org.eclipse.jface.dialogs.MessageDialog;
+import org.eclipse.swt.widgets.Display;
+import org.osgi.service.prefs.BackingStoreException;
+
+public class CompilerPluginVersionsChecker {
+
+    private final String FILENAME_WITH_PROJECT_SCOPED_PREFERENCES = "com.arc.cdt";
+    private final String UCLIBC_PLUGIN_VERSION = "arc.uclibc_plugin.version";
+    private final String ELF_PLUGIN_VERSION = "arc.elf32_plugin.version";
+
+    // These version numbers should be incremented when incompatible changes appear in these
+    // plug-ins.
+    private final int UCLIBC_PLUGIN_VERSION_NUMBER = 1;
```

Comment:
It checks for both ELF and uClibc plugin versions, but it is included only in elf plugin. So if elf plugin verion advances, but uClibc not - uClibc projects would still fail to load, because they check for both versions. I'd need either two classes or an argument to this class, so it would separate uClibc projects from ELF. So I'm not against having a single class, since there is a lot shared code, but it should separate projects by toolchain.